internal/contracts/manager.go

```diff
@@ -811,7 +840,118 @@ func (cm *contractManager) checkContractListenerExists(ctx context.Context, list
 		database.ContractListenerQueryFactory.NewUpdate(ctx).Set("backendid", listener.BackendID))
 }
 
-func (cm *contractManager) AddContractListener(ctx context.Context, listener *core.ContractListenerInput) (output *core.ContractListener, err error) {
+func (cm *contractManager) parseContractListenerFilters(ctx context.Context, listener *core.ContractListenerInput) (err error) {
+	// Handle deprecated root event
+	if len(listener.Filters) == 0 {
+		// Copy the deprecated interface into the first element in the filters array
+		listener.Filters = append(listener.Filters, &core.ListenerFilterInput{
+			ListenerFilter: core.ListenerFilter{
+				Event:     listener.Event,
+				Location:  listener.Location,
+				Interface: listener.Interface,
+			},
+			EventPath: listener.EventPath,
+		})
+	}
+
+	duplicateMapChecker := map[string]string{}
+	for _, filter := range listener.Filters {
+		if filter.Event == nil {
+			if filter.EventPath == "" || filter.Interface == nil {
+				return i18n.NewError(ctx, coremsgs.MsgListenerNoEvent)
+			}
+			// Copy the event definition into the filter
+			if filter.Event, err = cm.resolveEvent(ctx, filter.Interface, filter.EventPath); err != nil {
+				return err
+			}
+		} else {
+			filter.Interface = nil
+		}
+
+		if err := cm.validateFFIEvent(ctx, &filter.Event.FFIEventDefinition); err != nil {
+			return err
+		}
+
+		filter.Signature = cm.blockchain.GenerateEventSignature(ctx, &filter.Event.FFIEventDefinition)
+
+		if filter.Location != nil {
+			if filter.Location, err = cm.blockchain.NormalizeContractLocation(ctx, blockchain.NormalizeListener, filter.Location); err != nil {
+				return err
+			}
+		}
+
+		// Have parsed a filter with the same signature?
+		// In the case of location being different
+		if duplicateMapChecker[filter.Signature] != "" {
+			// If this duplicate filter is looking at all locations it's a superset of the previous one
+			// or the previous filter was also looking at all locations then we are trying to add a subset
+			if filter.Location == nil || duplicateMapChecker[filter.Signature] == "*" {
+				return i18n.NewError(ctx, coremsgs.MsgDuplicateContractListenerFilterLocation)
+			}
+
+			location, err := cm.blockchain.StringifyContractLocation(ctx, filter.Location)
```

Comment:
I see you've distinguished between `NormalizeContractLocation` returning JSON in a deterministic way and `StringifyContractLocation`. I'm not sure that's necessary.

It does seem like your proposal below is correct that `GenerateEventSignature` should take the location (or have a separate `GenerateEventSignatureWithLocation` function if there's some other use of `GenerateEventSignature` that needs to preserve the existing behavior).